### PR TITLE
[Handshake] Add basic System Verilog integration test.

### PR DESCRIPTION
**Pull request by mikeurbach**
This tests from the Standard dialect all the way to System Verilog
using Verilator. A simple function in the Standard dialect is given
and tested in a simple C++ script.

### CODE REVIEW

**Comment by mikeurbach** [comment]
As discussed in https://llvm.discourse.group/t/handshaketofirrtl-verifications/2555, we want to create some integration tests. This gets the ball rolling with a very basic example that can be lowered into a purely combinational circuit. Up next would be some more complex tests to flex the more complex cases like branching and merging, looping, memories, etc.

**Comment by mikeurbach** [comment]
Thanks @teqdruid for all the effort you put in to standing up this testing infrastructure. It is really nice to be able to do this and it just works.

File: integration_test/Dialect/Handshake/basic.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Vtop.h"
+
+int main(int argc, char **argv) {
```
Comment (lattner) [inline integration_test/Dialect/Handshake/basic.cpp:11]: Why does this need its own verilator driver?  I thought the rtl-sim script had one?

Comment (teqdruid) [reply integration_test/Dialect/Handshake/basic.cpp:11]: This is what I was saying during that simulator discussion we were having. This driver is necessary to provide stimulus to the design.

File: integration_test/Dialect/Handshake/basic.mlir [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// REQUIRES: rtl-sim
```
Comment (teqdruid) [inline integration_test/Dialect/Handshake/basic.mlir:1]: Since this test is only designed for verilator, it should require "verilator".

File: integration_test/Dialect/Handshake/basic.mlir [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+// RUN: circt-opt %s -create-dataflow -canonicalize-dataflow -canonicalize -cse > %t0.mlir
+// RUN: circt-opt %t0.mlir -lower-handshake-to-firrtl -canonicalize -cse > %t1.mlir
+// RUN: circt-opt %t1.mlir -pass-pipeline='firrtl.circuit(firrtl.module(firrtl-lower-types))' > %t2.mlir
+// RUN: circt-translate %t2.mlir -emit-firrtl-verilog > %t3.sv
+// RUN: circt-rtl-sim.py %t3.sv %S/basic.cpp --no-default-driver
```
Comment (teqdruid) [inline integration_test/Dialect/Handshake/basic.mlir:6]: And you'll need `--sim verilator` here to ensure that verilator is used. I'm working on a patch (slowly) to allow users to select the simulator they want to run for all the integration tests.

**Review by teqdruid — APPROVED** [review]
Valid/ready semantics! This is an ideal fit for ESI and cosim! (But I think we should have separate integration tests nonetheless.)

Congrats to you and @hanchenye for getting this basic test working!

File: integration_test/Dialect/Handshake/basic.cpp (hunk 1 above)
Comment (lattner) [reply integration_test/Dialect/Handshake/basic.cpp:11]: It's it possible to declarativize this like we do for the other testing tools?  I don't see how this is different.

Comment (teqdruid) [reply integration_test/Dialect/Handshake/basic.cpp:11]: See how the stimulus (and checking) is different from "real" simulators? From a 10k foot view, it's only different in the language -- C++ vs SystemVerilog.

Making a stimulus and checking language which compiles to either C++ or SystemVerilog is something I also brought up in that discussion. I think there's a big opportunity to create a 'verification' dialect to do exactly that! But it'd be a big project without any input language. Starting with a simple dialect to fill our modest needs is probably a good start.

Either way, I think this is a quite a large ask.

**Comment by mikeurbach** [comment]
Yep @teqdruid there is definitely a good fit between Handshake and ESI. Every Handshake component has valid/ready semantics, and I'm not sure if we want ESI channels between every component. From my understanding, it would make sense to plug in ESI channels at `handshake.func` level of granularity. This is something I'm interested in exploring/prototyping.

File: integration_test/Dialect/Handshake/basic.cpp (hunk 1 above)
Comment (mikeurbach) [reply integration_test/Dialect/Handshake/basic.cpp:11]: A verification dialect sounds really interesting, but out of the scope for what I'm trying to achieve here.

Is this the correct way to write a Verilator test case? I've written testbenches in Verilog that instantiate the device under test, provide some stimuli, and print/assert on the DUT outputs. My understanding of Verilator is we are doing the same thing here, but with the driver in C++.

Comment (teqdruid) [reply integration_test/Dialect/Handshake/basic.cpp:11]: Frankly, I have no idea. I’m not a verification expert so I only write basic, ad-hoc SV tests. I’ve mostly used Verilator to connect up to larger pieces of software, not driven signals individually.

**Comment by teqdruid** [comment]
Yes, ESI is intended for coarser grained connectivity. I don’t know enough about handshake in practice to say if func is a good granularity.

**Comment by lattner** [comment]
I would really prefer to see the C++ drivers eventually go away.  Such things are going to be necessarily verilator specific.  Such an approach is fine for a few tests, but isn't something that makes sense to scale.

**Comment by mikeurbach** [comment]
@lattner I see your point about making this verilator specific, and I definitely hope to add some integration tests here that will be good for the long term. Just to check my understanding, is the scalable route to write these test cases using System Verilog testbenches? I honestly don't know how people prefer to drive these kinds of tests, verilator was one idea that was thrown out.

**Comment by teqdruid** [comment]
Chris is saying that all integration tests should use ESI cosim to drive them. 😜

**Comment by mikeurbach** [comment]
I'm definitely interested in using in-tree tooling like ESI cosim and llhd-sim. I'm also interested what third-party simulators people use and how to write testbenches for them.

Perhaps we can discuss the topic of integration testing CIRCT on Wednesday?

**Review by stephenneuendorffer — COMMENTED** [review]
Indeed, this is very cool.  My thinking has been that there's actually quite alot of complexity in testing... to the extent that we probably need not only a verification dialect, but probably a library of transactor operations as well, along with transformations that can manipulate the testbenches, to leverage these transactors.  I'd be happy to give an overview of how Vivado HLS does cosimulation in this respect: there are quite a few nasty bits that I hope we can simplify in CIRCT.

**Comment by mikeurbach** [comment]
Thanks for presenting on cosim last week @stephenneuendorffer. Seems like there is a lot of really interesting work to be done in this area.

For the larger group that has weighed in here: I'd like to separate the discussions about cosim and the need to integration test the Handshake pipeline. As we discussed [on Discourse](https://llvm.discourse.group/t/handshaketofirrtl-verifications/2555), it would be good to have some tests for the Handshake pipeline's outputs at the HDL level. I think we really want to have such tests sooner rather than later, and we should be using a third-party simulator so we can focus on verifying the functionality, not finding edge cases in our own simulation tools. Are we in agreement here?

If so, I'd really like to try and come to consensus about how to add these tests. I picked Verilator for the third-party simulator because it was suggested in the original Discourse discussion, but comments here have pointed out that this means the test driver is inherently Verilator-specific, and Verilator may not even be the right tool for the job.

Perhaps we should return to the forum, but I'll ask here: if Verilator isn't the preferred way to integration test the Handshake lowering, what is?

**Comment by mikeurbach** [comment]
Hey all, I've just refreshed this PR with a completely different approach. This is now adding a System Verilog test driver, which is specific to Handshake, but should work generically for any Handshake test that returns an `i32`. It waits for the output to become valid, and then prints it for FileCheck to consume.

To seed the test cases, I borrowed some programs similar to the handshake-runner tests for [simple_loop](https://github.com/llvm/circt/blob/main/test/handshake-runner/simple_loop.mlir) and [loadstore](https://github.com/llvm/circt/blob/main/test/handshake-runner/loadstore.mlir).

The loop test relies on #602, so I hope we can converge on that one and land it. I've actually used this test driver to help me debug and implement #602, and I'm also planning on using the memory test to help me navigate the upcoming changes related to FIRRTL memories. Eventually, I would hope to have a more complex example, like [matmul](https://github.com/llvm/circt/blob/main/test/handshake-runner/matmul-check-std.mlir) or something similar.

The test driver should work with any IEEE simulator you have configured, but I've been testing locally with a free version of Vivado. It would be great if we could get a free version of a commercial simulator to enable this test in the nightly runs.

What do we think about the new approach?

**Comment by teqdruid** [comment]
> The test driver should work with any IEEE simulator you have configured, but I've been testing locally with a free version of Vivado. It would be great if we could get a free version of a commercial simulator to enable this test in the nightly runs.

I doesn't just have to be free (as in beer), the license has to allow usage in this way -- at a minimum, the license needs to allow re-distribution I think.

File: integration_test/Dialect/Handshake/driver.sv [hunk 4]
```diff
@@ -0,0 +1,33 @@
+module top();
+  logic clock = 0;
+  logic reset = 0;
+  logic arg0_valid, arg0_ready;
+  logic arg1_valid, arg1_ready;
+  logic [31:0] arg1_data;
+  logic arg2_valid, arg2_ready;
+
+  main main (.*);
```
Comment (teqdruid) [inline integration_test/Dialect/Handshake/driver.sv:9]: nit: This is typically called the 'dut' (device under test). I've been calling the dut 'top' and the driver 'driver'. It's a good rule-of-thumb to name the module after the filename. Concretely, I would re-name "top" to "driver" and "main" to "top".

**Review by teqdruid — APPROVED** [review]
LGTM! Thanks.

It's unfortunate this won't run in our nightly CI checks...

File: integration_test/Dialect/Handshake/driver.sv (hunk 4 above)
Comment (mikeurbach) [reply integration_test/Dialect/Handshake/driver.sv:9]: :+1: I started by copying std dialect code that was meant to work with mlir-cpu-runner, so this is coming from the C language world where `main` is the norm. I'll name these in the traditional hardware way.

**Review by lattner — APPROVED** [review]
LGTM.  If you're getting a pass pipeline established, I would recommend thinking about making a "firtool"-like driver that encapsulates this.

**Comment by mikeurbach** [comment]
Thanks for the reviews. The handshake pipeline is starting to shape up, will definitely think about a handshake-specific driver.

**Comment by lattner** [comment]
this patch lgtm, are you planning to land it?

**Comment by mikeurbach** [comment]
I was hoping to merge https://github.com/llvm/circt/pull/995 first to support the loop example and then this is good to go.
